src/messageParser.ts

```diff
@@ -176,139 +181,168 @@ function matchCurlyBrackets(value: string): XRegExp.MatchRecursiveValueNameMatch
 	});
 }
 
-export class Literal {
-	constructor(
-		public value: string,
-		public start: number,
-		public end: number
-	) { }
-
-	public toString = (): string => {
-		return `Literal(${this.value},${this.start},${this.end})`;
-	};
-
-	whereIs(offset: number): Literal | Message | null {
-		if (this.start < offset && offset < this.end) {
-			return this;
-		} else {
-			return null;
-		}
-	}
-}
-
 export class MessageList {
 	constructor(
 		public isReference: boolean,
 		public indentation: number,
-		public messages: Map<Key, Message>,
-		public metadata: Map<Key, Metadata>
+		public indentationCharacter: string,
+		public messageEntries: MessageEntry[],
+		public metadataEntries: MessageEntry[],
 	) { }
 
 	getPlaceholders(): Literal[] {
-		return Array.from(this.messages.values()).flatMap((message) => message.getPlaceholders());
+		return this.messageEntries.flatMap((messageEntry) => (messageEntry.message as Message).getPlaceholders());
+	}
+
+	getIndent(add?: number): string {
+		return this.indentationCharacter.repeat((this.indentation ?? 0) + (add ?? 0));
 	}
 
-	getMessageAt(offset: number): Message | Literal | Metadata | null {
-		const partsContaining = Array.from(this.messages.entries()).filter(([literal, message]) => literal.whereIs(offset) !== null || message.whereIs(offset) !== null);
+	getMessageAt(offset: number): Message | Metadata | null {
+		const partsContaining = [...this.messageEntries, ...this.metadataEntries]
+			.flatMap((entry) => [entry.key, entry.message])
+			.filter((value) => value.whereIs(offset) !== null);
 		if (partsContaining.length > 0) {
-			return partsContaining[0][0].whereIs(offset) ?? partsContaining[0][1].whereIs(offset);
+			return partsContaining[0].whereIs(offset);
 		} else {
 			return null;
 		}
```

Comment:
I think this might simplify slightly (removing the `if` statement) using `find` instead of `filter` which returns the first match (or undefined):

```js
.find((value) => value.whereIs(offset) !== null)
  ?.whereIs(offset);
```

Reply:
I was looking for something like `firstWhere`, thanks!